spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/channel/PostgresChannelMessageTableSubscriber.java

```diff
@@ -0,0 +1,161 @@
+
+import org.postgresql.PGNotification;
+import org.postgresql.jdbc.PgConnection;
+import org.springframework.context.SmartLifecycle;
+import org.springframework.core.log.LogAccessor;
+import org.springframework.integration.jdbc.store.JdbcChannelMessageStore;
+import org.springframework.integration.util.UUIDConverter;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+import java.sql.Statement;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class PostgresChannelMessageTableSubscriber implements SmartLifecycle {
+
+	private final LogAccessor logger = new LogAccessor(this.getClass()); // NOSONAR final
+
+	private final PgConnectionSupplier connectionSupplier;
+
+	private final String tablePrefix;
+
+	@Nullable
+	private Executor executor;
+
+	private final AtomicBoolean running = new AtomicBoolean();
+
+	private volatile Object identity;
+
+	private final Map<String, Set<PostgresChannelMessageTableSubscription>> subscriptions = new ConcurrentHashMap<>();
+
+	public PostgresChannelMessageTableSubscriber(PgConnectionSupplier connectionSupplier) {
+		this(connectionSupplier, JdbcChannelMessageStore.DEFAULT_TABLE_PREFIX);
+	}
+
+	public PostgresChannelMessageTableSubscriber(PgConnectionSupplier connectionSupplier, String tablePrefix) {
+		Assert.notNull(connectionSupplier, "A connectionSupplier must be provided.");
+		Assert.notNull(tablePrefix, "A table prefix must be set.");
+		this.connectionSupplier = connectionSupplier;
+		this.tablePrefix = tablePrefix;
+	}
+
+	public void setExecutor(Executor executor) {
+		this.executor = executor;
+	}
+
+	public boolean subscribe(PostgresChannelMessageTableSubscription subscription) {
+		Set<PostgresChannelMessageTableSubscription> subscriptions = this.subscriptions.computeIfAbsent(getKey(subscription.getGroupId()) + subscription.getRegion(), ignored -> ConcurrentHashMap.newKeySet());
+		return subscriptions.add(subscription);
+	}
+
+	public boolean unsubscribe(PostgresChannelMessageTableSubscription subscription) {
+		Set<PostgresChannelMessageTableSubscription> subscriptions = this.subscriptions.get(getKey(subscription.getGroupId()) + subscription.getRegion());
+		return subscriptions != null && subscriptions.remove(subscription);
+	}
+
+	@Override
+	public void start() {
+		if (running.getAndSet(true)) {
+			return;
+		}
+		Executor executor = this.executor;
+		if (executor == null) {
+			executor = Executors.newSingleThreadExecutor(job -> {
+				Thread t = new Thread(job);
+				t.setDaemon(true);
+				t.setName("postgres-channel-message-table-subscriber");
+				return t;
+			});
+			this.executor = executor;
+		}
+		// Avoid that stop/start sequence reactivates previously stopped thread.
+		Object current = new Object();
+		identity = current;
```

Comment:
You just avoid this via making those `start()` and `stop()` as `synchronized`.

Reply:
That would not work as the method returns before the executor thread might have ended. If the executor's thread dangles around, it could pick up a subsequent start. By switching the identity, it is certain that this additional listener should no longer listen or produce events, even if it takes the thread some time to complete.

Reply:
We probably need to wait in the `stop()` until that thread is done with its loop before returning from the method. Something like `CountDownLatch`

Reply:
I changed that to use a latch now.